lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -0,0 +1,408 @@
+//===- CheckCombLoops.cpp - FIRRTL check combinational cycles --*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the FIRRTL combintational cycles detection pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLInstanceGraph.h"
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "llvm/ADT/DepthFirstIterator.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/SCCIterator.h"
+#include "llvm/ADT/SmallSet.h"
+#include <variant>
+
+#define DEBUG_TYPE "check-comb-loops"
+
+using namespace circt;
+using namespace firrtl;
+struct Node {
+  Value val = nullptr;
+  size_t fieldId = 0;
+  Value leafVal = nullptr;
+  Node() = default;
+  Node(Value val, size_t fieldId, Value leafVal = nullptr)
+      : val(val), fieldId(fieldId), leafVal(leafVal) {}
+  Node(const Node &rhs) = default;
+  Node &operator=(const Node &rhs) = default;
+
+  bool operator==(const Node &rhs) const {
+    return val == rhs.val && fieldId == rhs.fieldId;
+  }
+  bool operator!=(const Node &rhs) const { return !(*this == rhs); }
+  bool isValid() { return val != nullptr; }
+  FModuleOp getParentModule() {
+    assert(val && "Must be a valid node");
+    if (val.isa<BlockArgument>())
+      return cast<FModuleOp>(val.getParentBlock()->getParentOp());
+    return val.getDefiningOp()->getParentOfType<FModuleOp>();
+  }
+};
+//
+//===----------------------------------------------------------------------===//
+// DenseMapInfo on Node
+//===----------------------------------------------------------------------===//
+
+namespace llvm {
+template <>
+struct DenseMapInfo<Node> {
+  static Node getEmptyKey() {
+    auto *pointer = llvm::DenseMapInfo<void *>::getEmptyKey();
+    return Node(Value::getFromOpaquePointer(pointer), 0);
+  }
+  static Node getTombstoneKey() {
+    auto *pointer = llvm::DenseMapInfo<void *>::getTombstoneKey();
+    return Node(Value::getFromOpaquePointer(pointer), 0);
+  }
+
+  static unsigned getHashValue(const Node &node) {
+    return mlir::hash_value(node.val);
+  }
```

Comment:
Is it fine to ignore fieldID?

Reply:
Updated to combine the hash value of fieldId.